freqtrade/data/entryexitanalysis.py

```diff
@@ -0,0 +1,255 @@
+
+
+def _analyze_candles_and_indicators(pair, trades, signal_candles):
+    buyf = signal_candles
+
+    if len(buyf) > 0:
+        buyf = buyf.set_index('date', drop=False)
+        trades_red = trades.loc[trades['pair'] == pair].copy()
+
+        trades_inds = pd.DataFrame()
+
+        if trades_red.shape[0] > 0 and buyf.shape[0] > 0:
+            for t, v in trades_red.open_date.items():
+                allinds = buyf.loc[(buyf['date'] < v)]
+                if allinds.shape[0] > 0:
+                    tmp_inds = allinds.iloc[[-1]]
+
+                    trades_red.loc[t, 'signal_date'] = tmp_inds['date'].values[0]
+                    trades_red.loc[t, 'enter_reason'] = trades_red.loc[t, 'enter_tag']
+                    tmp_inds.index.rename('signal_date', inplace=True)
+                    trades_inds = pd.concat([trades_inds, tmp_inds])
+
+            if 'signal_date' in trades_red:
+                trades_red['signal_date'] = pd.to_datetime(trades_red['signal_date'], utc=True)
+                trades_red.set_index('signal_date', inplace=True)
+
+                try:
+                    trades_red = pd.merge(trades_red, trades_inds, on='signal_date', how='outer')
+                except Exception as e:
+                    print(e)
+        return trades_red
+    else:
+        return pd.DataFrame()
+
+
+def _do_group_table_output(bigdf, glist):
+    if "0" in glist:
+        wins = bigdf.loc[bigdf['profit_abs'] >= 0] \
+                    .groupby(['enter_reason']) \
+                    .agg({'profit_abs': ['sum']})
+
+        wins.columns = ['profit_abs_wins']
+        loss = bigdf.loc[bigdf['profit_abs'] < 0] \
+                    .groupby(['enter_reason']) \
+                    .agg({'profit_abs': ['sum']})
+        loss.columns = ['profit_abs_loss']
+
+        new = bigdf.groupby(['enter_reason']).agg({'profit_abs': [
+                                                   'count',
+                                                   lambda x: sum(x > 0),
+                                                   lambda x: sum(x <= 0)]})
+        new = pd.concat([new, wins, loss], axis=1).fillna(0)
+
+        new['profit_tot'] = new['profit_abs_wins'] - abs(new['profit_abs_loss'])
+        new['wl_ratio_pct'] = (new.iloc[:, 1] / new.iloc[:, 0] * 100).fillna(0)
+        new['avg_win'] = (new['profit_abs_wins'] / new.iloc[:, 1]).fillna(0)
+        new['avg_loss'] = (new['profit_abs_loss'] / new.iloc[:, 2]).fillna(0)
+
+        new.columns = ['total_num_buys', 'wins', 'losses', 'profit_abs_wins', 'profit_abs_loss',
+                       'profit_tot', 'wl_ratio_pct', 'avg_win', 'avg_loss']
+
+        sortcols = ['total_num_buys']
+
+        _print_table(new, sortcols, show_index=True)
+    if "1" in glist:
+        new = bigdf.groupby(['enter_reason']) \
+                   .agg({'profit_abs': ['count', 'sum', 'median', 'mean'],
+                         'profit_ratio': ['sum', 'median', 'mean']}
+                        ).reset_index()
```

Comment:
At a quick glance, it seems like all but "0" could be refactored / combined as 90% of the code (except for the groupby) is identical.
using the group-by as argument or variable (e.g. `x = ['enter_reason']`) - using that as groupby (`...groupby(x).`) - as well as for the columns `new.columns = x + ['num_buys', ...]`).

The first instance (`0`) seems slightly different - but for the other 4, i think this can reduce the code to almost 1/3rd of the current code.

Reply:
aaaaand done